testing/base/source_gen.cpp

```diff
@@ -22,6 +22,237 @@ auto SourceGen::Global() -> SourceGen& {
+    type_weight_sum += fixed_type_weight.weight;
+    // Add all the fixed type spellings as immediately valid.
+    valid_type_names_.Insert(gen.IsCpp() ? fixed_type_weight.cpp_spelling
+                                         : fixed_type_weight.carbon_spelling);
+  }
+
+  // Compute the number of declared types used. We expect to have a decent
+  // number of repeated names, so we repeatedly append the entire sequence of
+  // class names until there is some remainder of names needed.
+  int num_declared_types =
+      num_types * type_use_params.declared_types_weight / type_weight_sum;
+  for ([[maybe_unused]] auto _ : llvm::seq(num_declared_types / num_classes)) {
+    type_names_.append(class_names_.begin(), class_names_.end());
+  }
+  // Now append the remainder number of class names. This is where the class
+  // names being un-shuffled is essential. We're going to have one extra
+  // reference to some fraction of the class names and we want that to be a
+  // stable fraction.
+  type_names_.append(class_names_.begin(),
+                     class_names_.begin() + (num_declared_types % num_classes));
+  CARBON_CHECK(static_cast<int>(type_names_.size()) == num_declared_types);
+
+  // Use each fixed type weight to append the expected number of copies of that
+  // type. This isn't exact however, so we'll both clamp the number of copies to
+  // not overshoot and may end up with a remainder.
+  for (const auto& fixed_type_weight : type_use_params.fixed_type_weights) {
+    // For very small numbers of types we may exhaust the fixed types. We don't
+    // try to scale the weights down for simplicity, we just cut off early.
+    if (static_cast<int>(type_names_.size()) == num_types) {
+      break;
+    }
+    int num_fixed_type =
+        std::min<int>(num_types - type_names_.size(),
+                      num_types * fixed_type_weight.weight / type_weight_sum);
+    type_names_.append(num_fixed_type, gen.IsCpp()
+                                           ? fixed_type_weight.cpp_spelling
+                                           : fixed_type_weight.carbon_spelling);
+  }
+
+  // If we need a tail of types to hit the exact number, simply round-robin
+  // through the fixed types without any weighting. With reasonably large
+  // numbers of types this won't distort the distribution in an interesting way
+  // and is simpler than trying to scale the distribution down.
+  for ([[maybe_unused]] auto _ :
+       llvm::seq(1 + (num_types - type_names_.size()) /
+                         type_use_params.fixed_type_weights.size())) {
+    for (const auto& fixed_type_weight :
+         llvm::ArrayRef(type_use_params.fixed_type_weights)
+             .take_front(num_types - type_names_.size())) {
+      type_names_.push_back(gen.IsCpp() ? fixed_type_weight.cpp_spelling
+                                        : fixed_type_weight.carbon_spelling);
+    }
+  }
+  CARBON_CHECK(static_cast<int>(type_names_.size()) == num_types);
```

Comment:
This code seems like it might be more complex and fragile than necessary?
```suggestion
  while (static_cast<int>(type_names_.size()) < num_types) {
    for (const auto& fixed_type_weight :
         llvm::ArrayRef(type_use_params.fixed_type_weights)
             .take_front(num_types - type_names_.size())) {
      type_names_.push_back(gen.IsCpp() ? fixed_type_weight.cpp_spelling
                                        : fixed_type_weight.carbon_spelling);
    }
  }
  CARBON_CHECK(static_cast<int>(type_names_.size()) == num_types);
```